test/distributed/_composable/test_composability/test_3d_composability.py

```diff
@@ -0,0 +1,757 @@
+            },
+            self._test_3d_composability,
+        )
+        self.logger.info("Finished job: 3D composability test")
+
+    def _test_3d_composability(self, dp_degree, tp_degree, pp_degree, float8, async_tp, schedule_class):
+        TRACE_BUFFER_SIZE = "TORCH_NCCL_TRACE_BUFFER_SIZE"
+        TRACE_FILE = "TORCH_NCCL_DEBUG_INFO_TEMP_FILE"
+        DUMP_ON_TIMEOUT = "TORCH_NCCL_DUMP_ON_TIMEOUT"
+        ASYNC_ERROR_HANDLING = "TORCH_NCCL_ASYNC_ERROR_HANDLING"
+        SKIP_CLEANUP = "3"
+
+        world_size = int(dp_degree * tp_degree * pp_degree)
+        parallel_dims = ParallelDims(
+            dp=dp_degree,
+            tp=tp_degree,
+            pp=pp_degree,
+            world_size=world_size,
+            enable_loss_parallel=False,
+            dp_type="fsdp",
+            float8=float8,
+            async_tp=async_tp,
+            schedule_class=schedule_class,
+        )
+        device = torch.device("cuda", index=dist.get_rank())
+        torch.cuda.set_device(device)
+
+        def _warn_overwrite_env(env, val):
+            if env in os.environ:
+                self.logger.warning("ENV will be overridden based on job config")
+            os.environ[env] = val
+
+        _warn_overwrite_env(ASYNC_ERROR_HANDLING, SKIP_CLEANUP)
+        _warn_overwrite_env(TRACE_BUFFER_SIZE, str(0))
+
+        os.environ["TORCH_NCCL_AVOID_RECORD_STREAMS"] = "1"
+        torch.manual_seed(0)
+        model_args = ModelArgs(dropout_p=0.0)
+        model = Transformer(model_args)
+        world_mesh = parallel_dims.build_mesh(device_type="cuda")
+
+        if parallel_dims.float8:
+            scale_for_fsdp = False
+            if parallel_dims.dp_enabled and parallel_dims.dp_type == "fsdp":
+                scale_for_fsdp = True
+            # a no-op hander if float8 is not enabled
+            float8_handler = Float8Handler(parallel_dims, scale_for_fsdp)
+            # swap to Float8Linear based on float8 configs
+            float8_handler.convert_to_float8_training(model)
+
+        if parallel_dims.dp_enabled:
+            dp_mesh = world_mesh["dp"]
+            dp_degree, dp_rank = dp_mesh.size(), dp_mesh.get_local_rank()
+        else:
+            dp_degree, dp_rank = 1, 0
+
+        def loss_fn(pred, labels):
+            return torch.nn.functional.cross_entropy(
+                pred.flatten(0, 1), labels.flatten(0, 1)
+            )
+
+        if parallel_dims.pp_enabled:
+            pp_mesh = world_mesh["pp"]
+            pp_schedule, model_parts = self._pipeline_llama(
+                model, pp_mesh, parallel_dims, device, model_args, loss_fn
+            )
+            for m in model_parts:
+                self._parallelize_llama(m, world_mesh, parallel_dims)
+                m.to_empty(device="cuda")
+                m.train()
+        else:
+            self._parallelize_llama(model, world_mesh, parallel_dims)
+            model.train()
+            model_parts = [model]
+
```

Comment:
is the next step to add forward/backward and optimizer step? seems we are testing model init right now